### PR TITLE
Extract state-holding `ConsumerFlux`

### PR DESCRIPTION
**Pull request by bsideup**
Now `DefaultKafkaReceiver` is holding a reference to `ConsumerFlux` (due to `doOnConsumer`)
and only responsible for creating it from `receive*()` methods.

Next step would be to deprecate `doOnConsumer` and make `DefaultKafkaReceiver` stateless.

### CODE REVIEW

File: src/main/java/reactor/kafka/receiver/internals/DefaultKafkaReceiver.java [hunk 1]
```diff
@@ -15,807 +15,123 @@
...
-        for (ConsumerRecord<K, V> r : records)
-            offsetBatch.updateOffset(new TopicPartition(r.topic(), r.partition()), r.offset());
-        return Flux.fromIterable(records)
-                   .concatWith(transactionManager.sendOffsets(offsetBatch.getAndClearOffsets().offsets(), receiverOptions.groupId()))
-                   .doAfterTerminate(() -> awaitingTransaction.set(false));
+                return transactionManager.begin()
+                    .then(Mono.fromRunnable(() -> consumerFlux.awaitingTransaction.getAndSet(true)))
+                    .thenMany(Flux.fromIterable(consumerRecords))
+                    .concatWith(transactionManager.sendOffsets(offsetBatch.getAndClearOffsets().offsets(), receiverOptions.groupId()))
+                    .doAfterTerminate(() -> consumerFlux.awaitingTransaction.set(false));
```
Comment (simonbasle) [inline src/main/java/reactor/kafka/receiver/internals/DefaultKafkaReceiver.java:113]: should this also detect cancellation by any chance ?

File: src/main/java/reactor/kafka/receiver/internals/DefaultKafkaReceiver.java [hunk 2]
```diff
@@ -15,807 +15,123 @@
...
     public Flux<Flux<ConsumerRecord<K, V>>> receiveExactlyOnce(TransactionManager transactionManager) {
-        Flux<ConsumerRecords<K, V>> flux = withDoOnRequest(createConsumerFlux(AckMode.EXACTLY_ONCE));
-        return  flux.map(consumerRecords -> transactionManager.begin()
-                                 .then(Mono.fromCallable(() -> awaitingTransaction.getAndSet(true)))
-                                 .thenMany(transactionalRecords(transactionManager, consumerRecords)))
-                                 .publishOn(transactionManager.scheduler());
-    }
+        ConsumerFlux<K, V> consumerFlux = createConsumerFlux(AckMode.EXACTLY_ONCE);
+        return consumerFlux
+            .doAfterTerminate(this::dispose)
```
Comment (simonbasle) [inline src/main/java/reactor/kafka/receiver/internals/DefaultKafkaReceiver.java:97]: you know the drill by now ;p

**Comment by bsideup** [comment]
Hey @simonbasle,

I did plan to rework the copy/pasted code, but as a follow up, so that it is focused on the changes, not the migration/code moving. Otherwise, I am concerned that we will miss something during the review of an already big PR and forget about fixing it later. WDYT?

**Comment by simonbasle** [comment]
> Hey @simonbasle,
> 
> I did plan to rework the copy/pasted code, but as a follow up, so that it is focused on the changes, not the migration/code moving. Otherwise, I am concerned that we will miss something during the review of an already big PR and forget about fixing it later. WDYT?

@bsideup no problem, just keep track of these suggestions somewhere then 👍

**Pull request merged**